[리액트 8기 이재윤] 6주차 과제 1

jaeyooniverse opened the pull request:
이번 주도 잘 부탁드립니다! 감사합니다!

jaeyooniverse committed:
container, page 파일 별로 분리하기 위한 폴더 생성 및 분류 작업

현재 src 한 폴더 안에 너무 많은 파일들이 존재하기 때문에 관리하는 데에 있어 어려움이 있음
따라서, container와 해당 파일에 대한 테스트를 모아둔 containers 폴더와 앞으로 작업할 pages
폴더를 생성함

jaeyooniverse committed:
라우팅 기능을 위한 react-router-dom 설치

React에서 다른 페이지로 이동하기 위해 react-router-dom이라는 라이브러리를 설치한다.
강의 환경과 동일한 환경에서 과제를 해결해보고자 react-router-dom v6이 아닌
v5로 설치한다.

jaeyooniverse committed:
About 페이지가 렌더되는지 테스트하라

AboutPage를 렌더하면 'About 페이지 입니다.'라는 문구를 렌더해해야 합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
AboutPage를 렌더하라

AboutPage를 렌더합니다. AboutPage는 About이라는 텍스트를 가진 h2 태그와
'About 페이지 입니다.'라는 일반 텍스트를 출력합니다.

jaeyooniverse committed:
NotFound 페이지가 렌더되는지 테스트하라

NotFound 페이지를 렌더하면 '요청하신 페이지를 찾을 수 없습니다'라는 문구를 출력해야 합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
NotFoundPage를 렌더하라

NotFoundPage를 렌더합니다. NotFoundPage는 '요청하신 페이지를 찾을 수 없습니다'라는
일반 텍스트를 출력합니다.

jaeyooniverse committed:
Home 페이지가 렌더되는지 테스트하라

HomePage를 렌더하면 'About'이라는 텍스트를 가진 링크와 'Restaurants'라는 텍스트를 가진 링크를 출력합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
About 링크, Restaurants 링크를 렌더하는 HomePage

HomePage를 렌더합니다. HomePage는 About 링크와 Restaurants 링크를 출력합니다.
About 링크를 입력하면 /about으로 이동하고 Restaurants 링크를 입력하면
/restaurants로 이동합니다.

jaeyooniverse committed:
regions, categories, restaurants를 정상적으로 출력하는지 테스트하라

RestaurantsPage는 RegionsContainer, CategoriesContainer, RestaurantsContainer를 출력해야 합니다.
이를 위해 테스트로 받은 정보를 제대로 출력하는지 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
RegionsContainer, CategoriesContainer, RestaurantsContainer를 렌더하는 RestaurantPage

RestaurantPage는 loadInitialDate로 데이터를 불러와 각각의 컨테이너에게 데이터를 보내줍니다.
Region 버튼과 Category 버튼틀이 존재하며 Region과 Category 버튼을 누르면 조건에 해당하는
식당을 출력합니다

jaeyooniverse committed:
라우팅을 정상적으로 진행하는지 테스트하라

이번 과제에서부터 App 컴포넌트는 라우팅관련 기능을 수행하므로 과거의 테스트 코드는 제거하고
App 컴포넌트는 주어진 path로 이동하는 작업을 해야합니다. 따라서, 주어진 path에 해당하는 정보를
정상적으로 출력하는 지를를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
라우팅 기능을 위한 App.jsx, index.jsx 수정

라우팅 기능을 위해 기존 App.jsx에서 react-router-dom의 Switch, Route와 Link를 이용하여
라우팅 관련 설정을 했습니다. App.jsx에서 BrowserRouter를 사용해도 되지만, indent를
줄이고자 index에 BrowserRotuer를 추가하여 Switch 밖에 Router를 넣어줘야 하는 문제를
해결했습니다

jaeyooniverse committed:
Switch 밖에는 Rotuer를 넣어주어야 한다
HomePage.test.jsx 테스트 코드를 실행하면 About과 Restaurants가 Link 태그로 되어있기 때문에
Link를 사용하려면 밖에 Router를 사용하라는 에러가 나게 되고 해당 에러를 해결하기 위해
renderHomePage 함수 HomePage 컴포넌트를 react-router-dom 테스트할 때 사용하는 MemoryRouter로
감싸 정상적으로 테스트를 진행할 수 있도록 수정하였습니다.

hannut91 commented:
커밋메시지 아주 좋습니다 ㅎㅎ

jaeyooniverse committed:
fetchRestaurantDetail API가 정상적으로 호출되는지 테스트하라

fetchRestaurantDetail API가 정상적으로 호출되는지 확인하기 위해 미리 저장해 둔 테스트 데이터와 일치해야 합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
식당 정보를 불러오기 위한 fetchRestaurantDetail API 추가

식당 정보를 불러오기 위해 fetchRestaurantDetail이라는 API를 추가했으며 정상적인
테스트 진행을 위해 __mocks__ 폴더에 있는 api.js에도 fetchRestaurantDetail이라는
mock 함수를 추가했습니다.

jaeyooniverse committed:
loadRestaurantDetail 액션이 정상적으로 동작되는지 테스트하라

loadRestaurantDetail은 restaurantId가 존재할 때에만 식당 정보를 불러와야 합니다.
restaurantId가 존재할 때에는 setRestaurantDetail 액션을 실행하며
restaurantId가 존재하지 않을 때에는 아무런 액션도 실행되지 않아야 합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
setRestaurantDetail 액션이 정상적으로 동작되는지 테스트하라

setRestaurantsDetail은 받은 식당 정보를 restaurantDetail이라는 initialState에 담아줘야 합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
식당 정보 기능을 위한 setRestaurantDetail, loadRestaurantDetail 액션 추가

restaurantId를 이용하여 식당 정보를 가져오는 loadRestaurantDetail,
가져온 식당 정보를 저장하는 setRestaurantDetail 액션을 추가했습니다.

jaeyooniverse committed:
식당 정보를 저장하기 위한 setRestaurantDetail 액션 리듀서에 추가

불러온 식당 정보를 저장하기 위해 setRestaurantDetail이라는 액션은 restaurantDetail을 payload로 받아
restaurantDetail에 저장합니다.

jaeyooniverse committed:
테스트 진행 시 사용하기 위한 restaurantDetail 추가

테스트 진행 시 비교할 정보를 위해 fixtures에 restaurantDetail이라는 객체를 추가했습니다.
해당 객체에는 실제 API로 불러오는 정보와 같은 형식의  데이터가 존재합니다.

jaeyooniverse committed:
식당 정보를 확인하기 위해 텍스트를 Link로 변경

식당 정보를 확인하기 위해 기존 식당 제목만 일반 텍스트로 나와있었던 것을 react-router-dom의
Link로 변경하여 식당 제목을 클릭하면 식당 정보 페이지로 이동할 수 있도록 만들었습니다.

jaeyooniverse committed:
react-router-dom을 사용하기 위한 index.html 수정

react-router-dom을 사용하기 전에 앞서 webpack.config.js에서 devServer 객체에 historyApiFallback
index:'index.html'이라는 프로퍼티를 추가해야 합니다. 따라서 index.html을 수정했습니다.

See also: https://github.com/bripkens/connect-history-api-fallback\#introduction

jaeyooniverse committed:
RestaurantDetailPage 라우팅이 제대로 되는지 테스트하라

RestaurantDetailPage로 라우팅이 제대로 되는지 테스트하기 위해 restaurantDetail에 정보가 담겨있다면
식당 정보 페이지에서 식당 이름과 주소, 메뉴 텍스트가 존재해야 합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
react-router-dom v6 업그레이드로 인한 마이그레이션

react-router-dom이 v5에서 v6로 업그레이드 되었고 강의에선 v5로 진행했지만 v6가 최신 버전이니
안정성을 위해 v6로 업데이트 했으며 이로 인한 변경 사항을 수정했습니다.

jaeyooniverse committed:
RestaurantDetail 페이지가 렌더되는지 테스트하라

식당에 대한 정보를 출력하는 RestaurantDetail 페이지는 fixtures에 restaurantDetail을 받아와
해당 데이터를 식당 이름, 주소, 메뉴 형식에 맞게 출력해야 합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
식당 정보를 출력하는 RestaurantDetailPage

RestaurantDetailPage는 react-router-dom의 useParams를 이용하여 restaurantId를 받아와
restaurantId에 해당하는 식당 정보를 출력합니다.

jaeyooniverse committed:
라우팅 기능 테스트를 위한 App.test.jsx, RestaurnatsPage.jsx 수정

/restaurants 로 이동하면 비동기로 불러온 서울 버튼이 존재하기 때문에 서울이라는 텍스트가
있는지 확인하려 했으나 비동기로 진행하기 때문에 test는 서울 버튼을 불러오기 전에 테스트를
진행해서 서울이라는 문구를 확인할 수 없으므로 fail을 반환한다. 따라서, App의 기능은
라우팅을 제대로 하는 지 테스트하는 컴포넌트이므로 RestaurantsPage.jsx에 Restaurants라는
h2 텍스트를 만들어서 Restaurants라는 문구가 존재하는 지 확인할 수 있으면 라우팅 기능
테스트는 진행할 수 있으므로 이처럼 확인할 수 있는 테스트로 수정했습니다.

jaeyooniverse committed:
react-router-dom 에러를 해결하기 위한 RestaurantContainer.test, RestaurantsPage.test 수정

react-router-dom에서 Link 태그를 사용하려면 밖에 Router 태그로 감싸줘야 한다.
따라서, Link 태그에 영향을 받는 RestaurantsContainer.test.jsx, RestaurantsPage.test.jsx에서
render할 때, react-router-dom을 테스트할 때 사용하는 MemoryRouter로 감싸줌으로써 해결했습니다.

jaeyooniverse committed:
restaurantDetail이 null일 경우 나오는 에러를 해결하기 위한 수정

restaurantDetail은 useParams로 얻은 restaurantId를 비동기로 loadResturantDetail을 진행하고
setRestaurantDetail을 통해 받은 정보를 useSelector로 얻어오는 과정에서 시간 소요가 되기 때문에
이러한 작업이 끝나기 전에 페이지가 렌더되면 restaurantDetail이 null이라는 에러를 만나게 된다.
따라서 restaurantDetail이 null인 경우 레스토랑 정보가 없어요 라는 문구를 출력하고 restaurantDetail이
존재하면 해당 정보를 출력할 수 있도록 수정한다.

jaeyooniverse committed:
coverage 100%을 위한 RestaurantsDetailPage.test.jsx 수정

앞서, RestaurantsDetailPage가 변경됐기 때문에 restaurantDetail이 0일 경우 테스트 코드를 작성해야
coverage 100%에 도달할 수 있다. 따라서, useSelector를 이용하여 restaurantDetail이 존재하는 경우,
존재하지 않는 경우로 나눠 존재하는 경우는 식당 정보를 확인하는 테스트를, 존재하지 않는 경우는
'레스토랑 정보가 없어요'라는 문구를 확인하는 테스트를 추가했습니다.

jaeyooniverse committed:
e2e 테스트 통과를 위한 AboutPage, AboutPage.test 수정

e2e 테스트에서 About 페이지로 이동했을 때 '404 Not Found'라는 텍스트 문구를 보기 원하므로
AboutPage에서 문구를 '404 Not Found'로 수정했으며 이와 같이 테스트 코드도 '404 Not Found'를
출력하는 테스트로 수정했습니다.

jaeyooniverse commented:
현재 통과하지 못하고 있는 e2e 테스트가 존재하는데, 이를 해결하지 못해서 질문드립니다.

코드숨 과제 데모 영상과 같이 /restaurants에서 식당 제목 링크를 눌러 이동하면 식당을 정상적으로 출력하지만, Home 페이지에서 /restautant/1 로 url을 바로 입력하여 이동하면 식당이 출력되지 않아 테스트를 통과하지 못하고 있습니다. console 창에서도 확인할 수 있는 에러는 404뿐이라 react-router-dom의 공식 문서와 구글링을 해보며 해결책을 찾아봐도 해결책을 얻지 못해서 어떻게 해결할 수 있을지 질문드립니다.

hannut91 commented:
```jsx
<Routes>
      <Route path="/" element={<HomePage />} />
      <Route path="/about" element={<AboutPage />} />
      <Route path="/restaurants" element={<RestaurantsPage />} />
      <Route
        path="/restaurants/:restaurantId"
        element={<RestaurantDetailPage />}
      />
      <Route path="/*" element={<NotFoundPage />} />
</Routes>
```

클릭했을 때는 `/restaurants/{id}`로 요청해야 할 것 같은데, url에 입력할 떄는 뒤에 s를 빠트려서 `/restaurant/1`로 요청해서 그런 것 같은데 url을 한 번 확인해 보세요

hannut91 commented on src/App.jsx:
```diff
@@ -1,32 +1,29 @@
+        <Route path="/about" element={<AboutPage />} />
+        <Route path="/restaurants" element={<RestaurantsPage />} />
+        <Route
+          path="/restaurants/:restaurantId"
```
앞에 `restaurants`가 있어서 `/restaurants/:id`라고만 써줘도 될 것 같아요

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,88 @@
+      regions: [],
+      categories: [],
+      restaurants: [],
+      restaurantDetail: {
```
만들어놓은 fixture를 사용하는 것은 어떨까요

hannut91 commented on src/pages/AboutPage.jsx:
```diff
@@ -0,0 +1,8 @@
+  return (
+    <div>
+      <h2>About</h2>
+      404 Not Found
```
About인데 404가 있네요

jaeyooniverse replied:
NotFoundPage로 수정해야 하는걸 AboutPage를 수정했네요 ㅠㅠ 수정했습니다
done: https://github.com/CodeSoom/react-week6-assignment-1/pull/87/commits/1b2e095b05b8570ba439d1e6ee8fd4a56c01a8cd

hannut91 commented on src/pages/RestaurantDetailPage.jsx:
```diff
@@ -0,0 +1,42 @@
+      </p>
+      <h3>메뉴</h3>
+      <ul>
+        {menuItems?.map(({ id, name: menu }) => (
```
menuItems의 기본값은 빈 배열일거에요. 빈 배열에 map을 하는 것은 아무런 효과가 없습니다. 따라서 `?`를 쓰지 않아야 됩니다

jaeyooniverse replied:
수정했습니다!
done: https://github.com/CodeSoom/react-week6-assignment-1/pull/87/commits/ab696de0fcdd34ffbe20cfe7e855d997643d8c08

hannut91 commented on webpack.config.js:
```diff
@@ -14,4 +14,9 @@ module.exports = {
   resolve: {
     extensions: ['.js', '.jsx'],
   },
+  devServer: {
+    historyApiFallback: {
+      index: 'index.html',
+    },
+  },
```
이 문서도 읽어 봅시다

https://github.com/bripkens/connect-history-api-fallback#introduction

jaeyooniverse committed:
NotFoundPage 내용을 AboutPage에 넣어줬던 실수 수정

NotFoundPage의 내용을 404 Not Found로 바꾸려 했는데 실수로 AboutPage에 바꿨습니다.
따라서, AboutPage의 내용은 About 페이지 입니다.로 다시 수정하면서 AboutPage.test.jsx도
이와 같이 수정하고, 원래 진행하려 했던 NotFoundPage도 수정하면서 NotFoundPage.test.jsx로
변경했습니다.

jaeyooniverse committed:
index.html에서 main.js의 경로 수정

현재 url로 /restaurants/1로 이동하면 main.js를 /restaurants/main.js로 찾는 이슈가 있어
main.js를 /부터 시작하는 상대 경로를 이용하여 찾을 수 있도록 index.html을 수정함

jaeyooniverse committed:
NotFound 페이지 수정에 따른 App.test.jsx 수정

NotFound 페이지의 문구가 변경되었으므로 이를 테스트하는 App.test.jsx도 수정한다.

jaeyooniverse committed:
menuItesm의 기본값은 빈 배열이므로 ?제거

menuItem이 존재하지 않을 경우 아예 없는 것이 아닌 빈 배열로 존재할 것이므로 빈 배열에 ?를 사용하는 것은
빈 배열에 map을 실행하게 된다. 따라서 ?를 제거한다

hannut91 commented:
```
NotFound 페이지 수정에 따른 App.test.jsx 수정

NotFound 페이지의 문구가 변경되었으므로 이를 테스트하는 App.test.jsx도 수정한다.
```

위 커밋은 다른 작업이 들어가 있는 것 같네요. 뭔가 Not Found문구에 대한 것만 변경이 있을거라고 기대했는데 테스트용 데이터를 변경하고 있어요. 이런 경우 커밋을 2개로 만드는게 좋습니다.

그리고 문구를 변경한 작업과 테스트를 변경한 작업은 서로 긴밀하게 연관이 있으므로 같은 작업에 넣었으면 더 좋았을 것 같네요

hannut91 commented on src/actions.js:
```diff
@@ -67,3 +72,14 @@ export function loadRestaurants() {
+      return;
+    }
+
+    const restaurantDetail = await fetchRestaurantDetail({ restaurantId });
```
서버에서 예상치 못한 문제로 데이터를 불러오지 못할 경우 어떻게 해야 할까요?

jaeyooniverse replied:
서버에서 예상치 못한 문제로 데이터를 불러오지 못할 경우를 대비하여
먼저, null값을 넣어 setRestaurantDetail action을 진행한 다음
데이터를 불러왔을 경우 기존 코드처럼 진행하도록 하였으며, 이에 따른 테스트 코드도 수정했습니다.
예외 케이스를 알려주셔서 감사합니다!
done: https://github.com/CodeSoom/react-week6-assignment-1/pull/87/commits/dab17ecaac87e41e11c08d89fb703449f1f50dd6

jaeyooniverse committed:
서버에서 데이터를 불러오지 못했을 경우를 처리하기 위한 actions와 actions.test 수정

서버에서 데이터를 처리하지 못했을 경우를 대비하여 setRestaurantDetail에 null 값을 넣어주는 액션을
먼저 진행하고, 데이터를 받아왔다면 받아온 정보를 restaurantDetail에 담아 setRestaurantDetail 액션을
또 한 번 진행합니다. 또한, 다음과 같은 기능의 수정 사항을 테스트 하기 위한 테스트 코드를 수정했습니다.

jaeyooniverse committed:
식당 정보를 불러오는 시간동안 식당 정보가 없다는 표현을 로딩 중으으로 수정

비동기로 식당 정보 데이터를 불러오는 데에 소요되는 시간동안 restaurantDetail은 null값이고,
데이터를 불러왔을 때, 화면을 다시 render하므로 식당 정보가 없어요!라는 표현보다 로딩 중이라는
표현이 더 적합한 거 같아 이를 수정했습니다.

jaeyooniverse committed:
RestaurantDetailPage 수정에 따른 테스트 코드 수정

RestaurantDetailPage에서 텍스트 수정이 있었으므로 그에 따라 텍스트 문구를 확인하는
테스트 코드 또한 수정했습니다.

hannut91 commented:
음 일단 에러가 무엇인지부터 공부해 봐야 할 것 같아요. 아래 내용을 다시 공부해 봅시다. 그리고 적용해 보세요

- https://ko.javascript.info/try-catch
- https://ko.javascript.info/promise-error-handling

jaeyooniverse committed:
에러가 발생할 경우의 처리를 위한 RestaurantDetailPage.jsx 수정

기존 코드에서 에러가 발생할 경우 흰 화면이 나와서 사용자가 에러를 인지하지 못하고 있습니다.
따라서, try catch문으로 기존 작업을 진행하다 에러가 발생하면 '존재하지 않는 식당이거나 데이터를
불러오는 과정에서 에러가 발생했습니다.'라는 문구를 출력합니다.

hannut91 commented:
아직 어려우신 것 같네요. 힌트를 좀 더 드리자면, 리액트는 우리 앱의 상태를 화면에 보여주는 기능을 하는 라이브러리입니다. 에러가 발생했다면, 에러가 발생했다는 상태가 될 것이고, 이러한 상태일 때 상태를 보여주도록 해보세요
